printing a optional reference to an interface gives a c error: expected struct or union but not 'struct _option_main__IGameObject_ptr *'

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

**V doctor:**

```
V full version: V 0.4.9 5c65e58.47c0ca8
OS: windows, Microsoft Windows 11 Pro v26100 64-bit
Processor: 24 cpus, 64bit, little endian, 

getwd: D:\vbug2
vexe: C:\v\v.exe
vexe mtime: 2024-12-23 16:00:58

vroot: OK, value: C:\v
VMODULES: OK, value: C:\Users\info\.vmodules
VTMP: OK, value: C:\Users\info\AppData\Local\Temp\v_0

Git version: git version 2.42.0.windows.2
Git vroot status: 0.4.9-15-g47c0ca8a-dirty
.git/config present: true

CC version: Error: 'cc' is not recognized as an internal or external command,

operable program or batch file.


emcc version: Error: 'emcc' is not recognized as an internal or external command,

operable program or batch file.


thirdparty/tcc status: thirdparty-windows-amd64 b425ac82
```

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg main.v`

```v
module main

interface IGameObject {
mut:
	name     string
	parent   ?&IGameObject
	children []&IGameObject
	add_child(mut o IGameObject)
}

@[heap]
struct GameObject implements IGameObject {
mut:
	name     string
	parent   ?&IGameObject
	children []&IGameObject
}

fn (mut gameobject GameObject) add_child(mut o IGameObject) {
	o.parent = gameobject
	gameobject.children << o
}

fn main() {
	mut v1 := &GameObject{
		name: 'v1'
	}
	mut v2 := &GameObject{
		name: 'v2'
	}
	v1.add_child(mut v2)
	if v1.children[0].parent != none {
 		eprintln("HERE ITS: ${v1.children[0].parent.name}")
	}
}
```

**What did you expect to see?**

no c error

**What did you see instead?**

```
================== C compilation error (from tcc): ==============
cc: C:/Users/info/AppData/Local/Temp/v_0/main.01JFT40CFCTQ78FN4BSWTE5HXR.tmp.c:3796: warning: implicit declaration of function 'tcc_backtrace'
cc: C:/Users/info/AppData/Local/Temp/v_0/main.01JFT40CFCTQ78FN4BSWTE5HXR.tmp.c:5548: error: expected struct or union but not 'struct _option_main__IGameObject_ptr *'
... (the original output was 3 lines long, and was truncated to 2 lines)
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

<sub><a href="https://huly.app/guest/vlang-66f40c4d-a476b54c67-771fdd?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzY5ODk5MGYxNDRmNTg1YWU0ZmExYTkiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctYWxleGFuZGVyLXZsYW5nLTY2ZjQwYzRkLWE0NzZiNTRjNjctNzcxZmRjIn0.YVdwrf859Of8dLYXksz0ailblWty6BYrHJlYyLITPP0">Huly&reg;: <b>V_0.6-21684</b></a></sub>

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -4208,6 +4208,8 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 	mut sum_type_dot := '.'
 	mut field_typ := ast.void_type
 	mut is_option_unwrap := false
+	mut is_dereferenced := node.expr is ast.SelectorExpr && node.expr.expr_type.is_ptr()
+		&& g.table.final_sym(node.expr.expr_type).kind in [.interface, .sum_type]
 	if f := g.table.find_field_with_embeds(sym, node.field_name) {
 		field_sym := g.table.sym(f.typ)
 		field_typ = f.typ
@@ -4374,8 +4376,8 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 	}
 	alias_to_ptr := sym.info is ast.Alias && sym.info.parent_type.is_ptr()
 	if field_is_opt
-		|| ((g.unwrap_generic(node.expr_type).is_ptr() || sym.kind == .chan || alias_to_ptr)
-		&& node.from_embed_types.len == 0)
+		|| (((!is_dereferenced && g.unwrap_generic(node.expr_type).is_ptr())
+		|| sym.kind == .chan || alias_to_ptr) && node.from_embed_types.len == 0)
 		|| (node.expr.is_as_cast() && g.inside_smartcast) {
 		g.write('->')
 	} else {
@@ -4389,7 +4391,12 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 	}
 	g.write(field_name)
 	if is_option_unwrap {
-		g.write('.data))')
+		if field_typ.is_ptr() {
+			g.write('->')
+		} else {
+			g.write('.')
+		}
+		g.write('data))')
 	}
 	if sum_type_deref_field != '' {
 		g.write('${sum_type_dot}${sum_type_deref_field})')

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -4209,6 +4209,7 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 	mut field_typ := ast.void_type
 	mut is_option_unwrap := false
 	mut is_dereferenced := node.expr is ast.SelectorExpr && node.expr.expr_type.is_ptr()
+		&& !node.expr.typ.is_ptr()
 		&& g.table.final_sym(node.expr.expr_type).kind in [.interface, .sum_type]
 	if f := g.table.find_field_with_embeds(sym, node.field_name) {
 		field_sym := g.table.sym(f.typ)
@@ -4391,7 +4392,7 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 	}
 	g.write(field_name)
 	if is_option_unwrap {
-		if field_typ.is_ptr() {
+		if field_typ.is_ptr() && g.table.final_sym(node.expr_type).kind in [.sum_type, .interface] {
 			g.write('->')
 		} else {
 			g.write('.')

--- a/vlib/v/tests/options/option_selector_interface_ptr_test.v
+++ b/vlib/v/tests/options/option_selector_interface_ptr_test.v
@@ -1,0 +1,35 @@
+module main
+
+interface IGameObject {
+mut:
+	name     string
+	parent   ?&IGameObject
+	children []&IGameObject
+	add_child(mut o IGameObject)
+}
+
+@[heap]
+struct GameObject implements IGameObject {
+mut:
+	name     string
+	parent   ?&IGameObject
+	children []&IGameObject
+}
+
+fn (mut gameobject GameObject) add_child(mut o IGameObject) {
+	o.parent = gameobject
+	gameobject.children << o
+}
+
+fn test_main() {
+	mut v1 := &GameObject{
+		name: 'v1'
+	}
+	mut v2 := &GameObject{
+		name: 'v2'
+	}
+	v1.add_child(mut v2)
+	if v1.children[0].parent != none {
+		assert '${v1.children[0].parent.name}' == 'v1'
+	}
+}

